pdns/bpf-filter.cc

```diff
@@ -470,6 +506,101 @@ void BPFFilter::unblock(const ComboAddress& addr)
   }
 }
 
+void BPFFilter::block(const Netmask& addr, BPFFilter::MatchAction action)
+{
+  CounterAndActionValue value;
+
+  int res = 0;
+  if (addr.isIPv4()) {
+    CIDR4 key(addr);
+    auto maps = d_maps.lock();
+    auto& map = maps->d_cidr4;
+    if (map.d_count >= map.d_config.d_maxItems) {
+      throw std::runtime_error("Table full when trying to block " + addr.toString());
+    }
+
+    res = bpf_lookup_elem(map.d_fd.getHandle(), &key, &value);
+    if (res != -1 && value.action == action) {
```

Comment:
Correct me if I'm wrong because I did not test, but I think we will get a match if we are trying to insert a more specific subnet? Like, if we already have "127.0.0.0/24" in the map and try to insert "127.0.0.0/25", it looks like it will throw?
I think it might be better to only refuse to insert if we have an exact match, so we would likely need to check that the key is exactly the same if `bpf_lookup_elem()` return a match. That way we can have one action for a subnet and a different action for a more specific subnet, for example.
Does that make sense to you?

Reply:
I think throw a error is ok because of `value.action == action`. This means the same processing behavior as previously existing rules.
`blockRange("127.0.0.0/24")` and `blockRange("127.0.0.0/25")` will get a error.
But `blockRange("127.0.0.0/24")` and `unblockRange("127.0.0.0/25")` is ok because of the difficult actions.

Reply:
You are right, different actions will indeed work. It does mean that we cannot have an entry for a more specific subnet with the same action. That might be fine because it does not really make sense to have the same action for a general subnet and a more specific one, but I wonder if we might have a problem with that later, especially if the block rules are inserted by a program.. I guess we can cross that bridge when we get to it.

Reply:
I may not understand what you mean. Do we need to modify this code to allow adding entries for the subnet with the same action?

Reply:
You are right. But I still don't think it's a good idea to always allow users to add item with the same action for subnet. 
XDP programs run in the kernel and are relatively sensitive to runtime. The more items in the map, the longer the XDP program will run when it search in the map. Sometimes users may just forget to block the subnet under a larger netmask.
Maybe we can add an optional bool parameter called `force` to blockRange.
`blockRange(Netmask [, force])`
Insert a extra item of the same action for the subnet will be accepted if and only if `force==true`.
What do you think of this idea?

Reply:
That sounds like a very good idea!